### ISSUE TITLE
Unable to create SSL/TLS secure channel

### ISSUE THREAD

**Issue by StuartBaker65** [dotnet/wcf#2498]
I'm using a number of third party web services and would like the ability to supply my key in some way to allow the call to succeed.  This is probably low level but would be a great feature to prevent having to go through the rigmarole of manually creating the proxy.

> An error occurred while attempting to find services at "https://my.secure/page.asmx". The request was aborted. Could not create SSL/TLS secure channel.

**Comment by mconnew** [comment]
You are a little unclear in your description about what it is you are wanting to do. So I'm going to try and cover all the bases. Presuming you start you code like this:
```csharp
BasicHttpsBinding binding = new BasicHttpsBinding(BasicHttpsSecurityMode.Transport);
binding.Security.Transport.ClientCredentialType = HttpClientCredentialType.None;
endpointAddress = new EndpointAddress(new Uri("http://myserver/MyService.svc"));
factory = new ChannelFactory<IService>(binding, endpointAddress);
```
Then you can continue depending on what your use case it:

**You want to accept any certificate from the server**
```csharp
factory.Credentials.ServiceCertificate.SslCertificateAuthentication = new X509ServiceCertificateAuthentication();
factory.Credentials.ServiceCertificate.SslCertificateAuthentication.CertificateValidationMode = X509CertificateValidationMode.None;
```

**You want to specify a specific certificate for the server**
```csharp
factory.Credentials.ServiceCertificate.Authentication.CertificateValidationMode = X509CertificateValidationMode.Custom;
factory.Credentials.ServiceCertificate.Authentication.CustomCertificateValidator = new MyCertificateValidator();

public class MyCertificateValidator : X509CertificateValidator
{
    public override void Validate(X509Certificate2 certificate)
    {
        // Check if the certificate provided by the service is what you want. If it isn't, throw
        if(!CheckIfExpectedCertificate(certificate))
        {
            throw new Exception("Not my certificate");
        }
    }
}
```

**Specify a client certificate to use**
For the initial setup, you will need to set `ClientCredentialType = HttpClientCredentialType.Certificate` instead of `None`.
```csharp
factory.Credentials.ClientCertificate.SetCertificate(
                StoreLocation.CurrentUser,
                StoreName.My,
                X509FindType.FindByThumbprint,
                "<my certificate thumbprint");
```

If this doesn't cover it, please provide more details.

**Comment by mconnew** [comment]
If the service is a WCF service, unless you consider your wsdl to be secret, you could use a different binding security configuration for the mex endpoint to remove the client certificate requirement for the wsdl only. If hosting in IIS, you would need to configure it to accept client certificate instead of require. This is the default.

**Comment by mconnew** [comment]
@StuartBaker65, I just want to clarify something. Are you using the "Add Web Reference" option found in the "Add Service Reference" advanced settings, or are you using the regular "Add Service Reference" or are you using WCF Connection Services extension?

**Comment by StuartBaker65** [comment]
The project is a .Net Core for which I can only add a "Connected Reference"  - this brings up the WCF Web Service extension. 

On a .Net Framework 4.7.1 project I *can* add a "Service Reference" and specifiy credentials to get through.

**Comment by mconnew** [comment]
I believe this scenario is supposed to work. Assigning the issue to @mlacouture to look into as this is an issue with the WCF Connected Service. @StuartBaker65, are you using the latest release of WCF Connected Services?

**Comment by zhenlan** [comment]
@StuartBaker65, does this blog post from @mlacouture help you?

https://blogs.msdn.microsoft.com/webdev/2017/06/28/wcf-web-service-reference-mex-authentication/

**Comment by zhenlan** [comment]
@StuartBaker65, please let us know if you have an update. Closing for now.

**Comment by StuartBaker65** [comment]
Happy for this so to be closed; It appears recent update of VS2017 to 15.5.7 has cleared/cured my problem without any intervention from me.

Thanks to all for suggestions and comments
